### PR TITLE
Feature/en 3776/add hot spots

### PR DESCRIPTION
**Pull request by SungjunApp**
ticket: https://pixlee.atlassian.net/browse/EN-3776

### Background
Since mobile SDKs mimic features of Widget and Lightbox, there're plenty of good features to be added to the SDKs. One of them is Hotspots.

## What this pr is about?
- If `showHotspots = true` is passed from the app-side to PXLPhotoProductView, SDK draws hotspots on PXLPhotoProductView similar to lightbox
- Tapping a hotspot scrolls the product list UI and navigate to the right product.
- Tapping the content hide and show all hotspots.
- always read original url to get content's width and height for calculating the positions of hotspots on the screen because bounding_box_products's x, y, with, height are generated based on the original content's width height from Control Panel.

### [DEMO] A new example to the demo app
- how to access the example: Launch the app > main screen > tap `Hotspots in PXLPhotoProductView` button

### [DATA] Added a new field to PXLPhoto
- boundingBoxProducts contains hotspots' positions on the content
```java
#!java
public class PXLPhoto implements Parcelable {
...
    public List<PXLBoundingBoxProduct> boundingBoxProducts;
...
}

public class PXLBoundingBoxProduct implements Parcelable {
    @Json(name = "product_id")
    public String productId;

    @Json(name = "x")
    @NullableInt
    public int x;

    @Json(name = "y")
    @NullableInt
    public int y;

    @Json(name = "width")
    @NullableInt
    public int width;

    @Json(name = "height")
    @NullableInt
    public int height;

    @Json(name = "aspect_ratio")
    @NullableDouble
    public double aspectRatio;
...
}
```

### [UI] Added a new option to PXLPhotoProductView to display hotspots
- showHotspots: pxlPhotoProductView.setContent(..) is added to PXLPhotoProductView
      - true: display hotspots. tapping on the content show/hide hotspots.
      - false: don't display hotspots.
- note that  hotspots are only visible on `Image`. it's not available on `video`.

- Code Example
  - To display hotspots, you need to set `showHotspots = true`. Please see the example codes below.
  ```kotlin
  #!kotlin
  pxlPhotoProductView
            .setContent(..
                showHotspots = true,
                ....)
  ```

# Discussion 
Some hotspots are located outside of the screen. This's not a bug. It's from the different screen ratios of mobile devices and the display options which are ImageScaleType's [FIT_CENTER, CENTER_CROP].
### What does ImageScaleType do?
- FIT_CENTER displays all areas of the image keeping the ratio of the content and fill the rest of the screen with a blurry image. 
- CENTER_CROP also keeps the ratio of the content. It fills the screen and crops the rest of the areas.
Situation 1. FIT_CENTER: displays hotspots in most cases. but it cannot display hotspots that are located outside of the screen area.
Situation 2. CENTER_CROP:  hotspots are located outside of the screen because the areas of the image where certain hotspots are supposed to be located are cropped.

Here's the example

### Example 1: Best case
|Control Panel|
|---------|
||

|.FIT_CENTER|.CENTER_CROP|
|---------|---------|
|||

### Example 2: in CENTER_CROP, one hotspot is off the screen.
|Control Panel|
|---------|
||

|.FIT_CENTER|.CENTER_CROP|
|---------|---------|
|||

### Example 3: in CENTER_CROP, one hotspot is located under the product list UI.
|Control Panel|
|---------|
||

|.FIT_CENTER|.CENTER_CROP|
|---------|---------|
|||

### Example 4: 
- FIT_CENTER: a hotspot located top left is partially off the screen.
- CENTER_CROP: the same hotspot is totally off the screen.

|Control Panel|
|---------|
||

|.FIT_CENTER|.CENTER_CROP|
|---------|---------|
|||

### CODE REVIEW

**Review by jchen123 — APPROVED** [review]
amazing

File: pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/widgets/PXLPhotoProductView.kt [hunk 1]
```diff
@@ -170,13 +203,115 @@ class PXLPhotoProductView : FrameLayout, LifecycleObserver {
...
+    private var hiddenHotspots = false
+    private var hotspotsJob: Job? = null
+    private fun addHotspots() {
+        // video does not have hotspots.
+        if (!useHotspots || photoInfo?.pxlPhoto?.isVideo == true) return
+
+        hotspotsJob?.cancel()
+        hotspotsJob = scope.launch {
+            withContext(Dispatchers.IO) {
+                // delay below for other elements like products, and the main content to be loaded with the maximum resources
```
Comment (jharrilim) [inline pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/widgets/PXLPhotoProductView.kt:221]: Are those views? And if so, do we have access to them here to check whether each one has called their `onCreate`? Also, what happens if they aren't loaded?

**Comment by jchen123** [comment]
question on the hotspots that are off the screen: if theyre on the edge, should we just not display them?

**Comment by SungjunApp** [comment]
To decide this, I'd like to get an opinion from @kj13ennett how do you think about this? this situation happens because trying to display hotspots on a limited space on mobile devices leads to missing some of them in the cases of the screenshots above. If we don't like losing any of this, I can drag them to the screen area.

File: pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/widgets/PXLPhotoProductView.kt (hunk 1 above)
Comment (SungjunApp) [reply pixleesdk/src/main/java/com/pixlee/pixleesdk/ui/widgets/PXLPhotoProductView.kt:221]: Are those activities/fragments?
- PXLPhotoProductView extends FrameLayout which is a view and can be imported from Activities and fragments.

do we have access to them here to check whether each one has called their onCreate
- yes we can using LifecycleObserver. Everything of PXLPhotoProductView can be used when or after onCreate of Activity/Fragment is called. The Lifecycle of Activity and Fragment is observed with this code.
```kotlin
#!kotlin
class PXLPhotoProductView : FrameLayout, LifecycleObserver {
    private fun initView(context: Context) {
        ....
        lifecycleOwner.lifecycle.addObserver(this)
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    fun playVideoOnStart() {
        playVideoOnResume()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    fun playVideoOnResume() {
        fireAnalyticsOpenLightbox()
        pxlPhotoView.playVideo()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    fun stopVideoOnPause() {
        pxlPhotoView.pauseVideo()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    fun stopVideoOnStop() {
        stopVideoOnPause()
    }

    private var isAnalyticsOpenLightboxFired = false

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    private fun fireAnalyticsOpenLightbox() {
    }
}
```

Also, what happens if they aren't loaded?
- If hotspots aren't loaded, PXLPhotoProductView will work without them. I can't think of any of the cases. I think the only way of not loading hotspots could be that the SDK user manually deletes the parent view(`v_hotspots`) of hotspots.

**Review by jharrilim — APPROVED** [review]

**Comment by SungjunApp** [comment]
after the discussion is done, this pr will be merged to production with possible modification based on the decision.

**Comment by SungjunApp** [comment]
@jchen123 Answer to Jeff's question, we display all hotspots although they are on the edge. They are drawn on the view layer but located outside of the screen by giving them minus x or y positions.

**Comment by kj13ennett** [comment]
> To decide this, I'd like to get an opinion from @kj13ennett how do you think about this? this situation happens because trying to display hotspots on a limited space on mobile devices leads to missing some of them in the cases of the screenshots above. If we don't like losing any of this, I can drag them to the screen area.

I feel like this is an edge case we don't have to worry about - if users are going to crop their photos out they should just turn off hotspots to avoid this case.  I don't think it makes sense for us to hide or move hotspots based on how a photo is cropped

**Comment by SungjunApp** [comment]
Thanks, @kj13ennett for your insight! 

I also think that not relocating hotspots off the screen makes sense since the relocation can lead to unexpected overlapping hotspots.  @jchen123 How do you think about this?

**Comment by jchen123** [comment]
Yes i agree

**Comment by SungjunApp** [comment]
fyi, I found that bounding_box_products does not filter based on the region_id. It has all information. To tackle it, I made a change that hotspots having its `[bounding_box_products]`.`product_id` matched with `products`.`id` are displayed.

**Comment by jchen123** [comment]
Yes, thats right.  Makes sense!

**Comment by jchen123** [comment]
does this add documentation to the readme somewhere?

**Comment by SungjunApp** [comment]
Thanks @jchen123. Good catch. I updated the README.

**Pull request merged**